### PR TITLE
101 oligo seq pipeline filter high abundance kmers

### LINKED ISSUE #101: [Oligo-Seq] filter high abundance kmers
- [ ]  filter high abundance kmers
    - automatic calculation of highly abundant k-mers? how do you define highly abundant?
    - use prohibited sequence filter to filter those kmers

- ProhibitedSequences:
    - filter high abundance kmers → automatic calculation of highly abundant k-mers? how do you define highly abundant?

### PR DESCRIPTION
**Pull request by lisa-sousa**
## Summary

This PR add kmer counting functionalities, refactors common preprocessing logic from pipeline files into reusable utility functions, improves FASTA and VCF file indexing handling, and adds comprehensive test coverage. These changes reduce code duplication, improve maintainability, and fix several issues related to index file management.

#### Pipeline Utility Functions (`oligo_designer_toolsuite/pipelines/_utils.py`)
- **`get_highly_abundant_kmer_sequences()`**: Refactored k-mer abundance counting and prohibited sequence generation into a reusable utility function that returns a list of prohibited sequences.

#### Sequence Processing Utilities (`oligo_designer_toolsuite/utils/_sequence_processor.py`)
- **`count_kmer_abundance()`**: New utility function for counting k-mer abundances across multiple FASTA files. Supports single k values, k ranges (tuple), or lists of k values. Returns fractional abundances sorted by abundance in descending order. This function is used by `get_highly_abundant_kmer_sequences()` and is available as a standalone utility for other use cases.

#### Pipeline Utility Functions (`oligo_designer_toolsuite/pipelines/_utils.py`)
- **`preprocess_tm_parameters()`**: Extracted melting temperature parameter preprocessing logic that was duplicated across multiple pipeline files. Converts string table names (e.g., "DNA_NN3") to `Bio.SeqUtils.MeltingTemp` objects.

This function are now used across all pipeline files (`_oligo_seq_probe_designer.py`, `_merfish_probe_designer.py`, `_scrinshot_probe_designer.py`, `_cycle_hcr_probe_designer.py`, `_seqfish_plus_probe_designer.py`), eliminating code duplication.

#### FASTA Indexing (`oligo_designer_toolsuite/utils/_sequence_parser.py`)
- **Added `index_fasta_file()` method to `FastaParser` class**: Explicitly creates or refreshes FASTA index files (`.fai`) using `samtools faidx`, ensuring compatibility with `bedtools getfasta` which relies on this tool.
- **Updated `write_database_to_file()` in `ReferenceDatabase`**: Now explicitly calls `index_fasta_file()` after writing FASTA databases to ensure index files are created.

#### VCF File Handling (`oligo_designer_toolsuite/utils/_sequence_parser.py`)
- **Refactored `merge_vcf_files()`**: Now removes all temporary files to ensuring proper cleanup. Only the final merged VCF file is written to the output directory.

#### Index File Cleanup (`oligo_designer_toolsuite/utils/_sequence_processor.py`, `oligo_designer_toolsuite/oligo_specificity_filter/_filter_base.py`)
- **Improved `remove_index_files()` function**: Simplified logic to remove only index files (files starting with `reference_file.`) while preserving the reference file itself.
- **Fixed `remove_reference()` in `_filter_base.py`**: Corrected regex pattern using `re.escape()` to properly match index files (`.fai`, `.csi`, `.tbi`, BLAST/Bowtie indexes).
- **Enhanced `load_database_from_file()`**: Now explicitly removes index files when `database_overwrite=True` to prevent stale index issues.

### 📦 Dependencies

- **Added `samtools >=1.22`** to installation requirements in `README.md` across all platform sections (Linux, MacOS M Chip, MacOS Intel Chip). This is required for FASTA indexing functionality.

#### New Test Classes (`tests/test_utils.py`)
- **`TestPreprocessTmParameters`**: Tests conversion of string table names to `Bio.SeqUtils.MeltingTemp` objects.
- **`TestGetHighlyAbundantKmerSequences`**: Tests k-mer abundance identification with various thresholds and file inputs.
- **`TestCountKmerAbundance`**: Comprehensive tests for the new `count_kmer_abundance()` function, including single k values, ranges, lists, multiple files, fractional abundance calculation, sorting, uppercase conversion, and edge cases (empty files, invalid inputs).
- **`TestFastaParserIndexFastaFile`**: Tests FASTA index file creation and stale index removal.
- **`TestRemoveIndexFiles`**: Tests removal of various index file types (`.fai`, `.csi`, `.tbi`, `.nhr`, `.nin`, `.nsq`, `.ebwt`, `.bt2`) while ensuring reference files are preserved.

#### Updated Tests (`tests/test_database.py`)
- **`test_write_database`**: Now verifies that `.fai` index files are created when writing FASTA databases.
- **`test_load_database_from_file`**: New test verifying that index files are properly removed when `database_overwrite=True` for both FASTA and VCF files.

## Breaking Changes

- `check_if_dna_sequence`now only accepts "A", "C", "T", "G", not "U" anymore.

Closes #101 

---

> [!NOTE]
> Introduces k-mer-driven prohibited sequence filtering and shared Tm preprocessing, overhauls FASTA/VCF index creation/cleanup, and updates CI/docs (adds samtools) with comprehensive tests.
> 
> - **Pipelines**:
>   - Oligo-Seq: integrate `ProhibitedSequenceFilter`; compute prohibited k-mers via `get_highly_abundant_kmer_sequences`; adopt `preprocess_tm_parameters`; streamline specificity param names; wire new filters.
>   - MERFISH, SCRINSHOT, SeqFISH+, CycleHCR: replace inline Tm table mapping with `preprocess_tm_parameters`.
> - **Configs**:
>   - `data/configs/oligo_seq_probe_designer.yaml`: add `target_probe_kmer_abundance_threshold`, `target_probe_prohibited_sequences`; unify specificity keys to generic `*_search_parameters`/`*_hit_parameters`.
> - **Specificity Filters**:
>   - Rename internals to `_create_reference`/`_remove_reference`; centralize index cleanup via `remove_index_files`; update BLAST/Bowtie/HybridizationProbability/Variants to new API.
> - **Database**:
>   - `ReferenceDatabase`: on overwrite, remove related index files; on write (FASTA) create `.fai` via `FastaParser.index_fasta_file`.
> - **Utils**:
>   - New: `count_kmer_abundance`, `remove_index_files` (robust multi-tool index deletion).
>   - `FastaParser.index_fasta_file` (samtools `faidx`); `VCFParser.merge_vcf_files` cleans temp/index artifacts.
>   - Pipelines utils: add `preprocess_tm_parameters`, `get_highly_abundant_kmer_sequences`.
>   - `check_if_dna_sequence`: simplify valid-char handling (no U warning path).
> - **CI/Docs**:
>   - Workflows (Ubuntu/macOS arm64/x64): install `samtools>=1.22`; minor CI config tweaks; include `cycle_hcr` config in sed step.
>   - README: require `samtools>=1.22`; generalize tool install wording.
> - **Tests**:
>   - Add coverage for Tm preprocessing, k-mer counting, high-abundance k-mers, index file creation/removal, and DB overwrite index cleanup; update property filter tests to `prohibited_sequences` arg.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit addcf8c69f6c1a067fe790740a0ff8636ec724be. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tests/test_oligo_property_filter.py [hunk 1]
```diff
@@ -126,14 +126,14 @@ def test_hardmasked_filter_Q(self):
 
 class TestSequenceContentFilters(unittest.TestCase):
 
-    def setUp(self):
-        self.prohibited_sequence_filter_str = ProhibitedSequenceFilter(prohibited_sequence="ACT")
-        self.prohibited_sequence_filter_list = ProhibitedSequenceFilter(prohibited_sequence=["ACT", "CCGC"])
+    def setUp(self) -> None:
+        self.prohibited_sequence_filter_str = ProhibitedSequenceFilter(prohibited_sequences="ACT")
+        self.prohibited_sequence_filter_list = ProhibitedSequenceFilter(prohibited_sequences=["ACT", "CCGC"])
```
[the remark this answers is not in the document]
Comment (lisa-sousa) [reply tests/test_oligo_property_filter.py:131]: Done

File: oligo_designer_toolsuite/utils/_sequence_processor.py [hunk 2]
```diff
@@ -124,3 +126,146 @@ def append_nucleotide_to_sequences(input_fasta: str, nucleotide: str) -> str:
...
+            raise ValueError("List of k values cannot be empty")
+        k_values = k
+    else:
+        raise ValueError(f"k must be int, tuple[int, int], or list[int], got {type(k)}")
+
+    if any(not isinstance(ki, int) or ki < 1 for ki in k_values):
+        raise ValueError("All k values must be positive integers")
+
+    # Normalize files_fasta input to a list
+    files_fasta = check_if_list(files_fasta)
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/utils/_sequence_processor.py]: cast_to_list

File: oligo_designer_toolsuite/utils/_sequence_processor.py [hunk 3]
```diff
@@ -124,3 +126,146 @@ def append_nucleotide_to_sequences(input_fasta: str, nucleotide: str) -> str:
...
+                file_path = os.path.join(root, file)
+                os.remove(file_path)
+
+
+def count_kmer_abundance(
+    files_fasta: str | list[str],
+    k: int | tuple[int, int] | list[int],
+) -> dict[int, dict[str, float]]:
+    """
+    Counts k-mer abundances across multiple FASTA files and returns fractional abundances.
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/utils/_sequence_processor.py:176]: maybe add documentation that the kmer abundance is calculated across all FASTA files and not per file

File: oligo_designer_toolsuite/pipelines/_oligo_seq_probe_designer.py [hunk 4]
```diff
@@ -728,6 +734,7 @@ def filter_by_property(
         Tm_chem_correction_parameters: dict | None,
         Tm_salt_correction_parameters: dict | None,
         homopolymeric_base_n: dict[str, int],
+        prohibited_sequences: list[str],
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/pipelines/_oligo_seq_probe_designer.py:737]: docstring missing

File: oligo_designer_toolsuite/pipelines/_scrinshot_probe_designer.py [hunk 5]
```diff
@@ -2025,18 +2025,8 @@ def main() -> None:
...
-    target_probe_Tm_parameters["tmm_table"] = getattr(mt, target_probe_Tm_parameters["tmm_table"])
-    target_probe_Tm_parameters["imm_table"] = getattr(mt, target_probe_Tm_parameters["imm_table"])
-    target_probe_Tm_parameters["de_table"] = getattr(mt, target_probe_Tm_parameters["de_table"])
-
-    detection_oligo_Tm_parameters = config["detection_oligo_Tm_parameters"]
-    detection_oligo_Tm_parameters["nn_table"] = getattr(mt, detection_oligo_Tm_parameters["nn_table"])
-    detection_oligo_Tm_parameters["tmm_table"] = getattr(mt, detection_oligo_Tm_parameters["tmm_table"])
-    detection_oligo_Tm_parameters["imm_table"] = getattr(mt, detection_oligo_Tm_parameters["imm_table"])
-    detection_oligo_Tm_parameters["de_table"] = getattr(mt, detection_oligo_Tm_parameters["de_table"])
+    ##### Preprocess Tm parameters #####
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/pipelines/_scrinshot_probe_designer.py:2028]: minor: style not consistent with other pipelines

File: oligo_designer_toolsuite/pipelines/_utils.py [hunk 6]
```diff
@@ -264,3 +267,64 @@ def format_sequence(database: OligoDatabase, property: str, region_id: str, olig
     if not isinstance(value, str):
         raise ValueError(f"Expected string for {property}, got {type(value)}")
     return value
+
+
+def preprocess_tm_parameters(tm_parameters: dict[str, Any]) -> dict[str, Any]:
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/pipelines/_utils.py:272]: maybe cleaner not to modify in place but return a changed copy? maybe a bit easier to reason about the function how it is used now; also for future pydantic model use could be easier?

File: oligo_designer_toolsuite/utils/_checkers_and_helpers.py [hunk 7]
```diff
@@ -68,8 +67,6 @@ def check_if_dna_sequence(seq: str, valid_characters: list[str] | None = None) -
         raise ConfigurationError(f"Valid characters must be single characters. Received: {valid_characters}.")
 
     valid_characters_upper = [char.upper() for char in valid_characters]
-    if not all(char.upper() in ["A", "C", "T", "G", "U"] for char in valid_characters_upper):
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/utils/_checkers_and_helpers.py:71]: something for NEWS.md?

File: oligo_designer_toolsuite/utils/_sequence_parser.py [hunk 8]
```diff
@@ -506,6 +506,41 @@ def merge_fasta_files(self, files_in: list[str], file_out: str, overwrite: bool
                 for idx in seq_record:
                     SeqIO.write(seq_record[idx], handle_fasta, "fasta")
 
+    def index_fasta_file(self, file_fasta: str) -> None:
+        """
+        Creates or refreshes the FASTA index file (.fai) for the given FASTA file.
+
+        This method ensures that a valid index exists for the FASTA file. If force_reindex
```
Comment (jonas-hag) [inline oligo_designer_toolsuite/utils/_sequence_parser.py]: what does force_reindex reference?

**Review by jonas-hag — CHANGES_REQUESTED** [review]
looks good; I added only some minor commens

File: oligo_designer_toolsuite/pipelines/_utils.py (hunk 6 above)
Comment (lisa-sousa) [reply oligo_designer_toolsuite/pipelines/_utils.py:272]: but th eobject is not usable without the prprocessing. That's why I modify in place. We can change i later though if required for pydantic

File: oligo_designer_toolsuite/utils/_sequence_parser.py (hunk 8 above)
Comment (lisa-sousa) [reply oligo_designer_toolsuite/utils/_sequence_parser.py]: referred to an old parameter that was removed. Doc string is adapted now.

File: oligo_designer_toolsuite/pipelines/_scrinshot_probe_designer.py (hunk 5 above)
Comment (lisa-sousa) [reply oligo_designer_toolsuite/pipelines/_scrinshot_probe_designer.py:2028]: yes, because we use it twice as parameter, so need to preprocess it twice

File: oligo_designer_toolsuite/pipelines/_oligo_seq_probe_designer.py [hunk 9]
```diff
@@ -1366,65 +1380,24 @@ def main() -> None:
...
-    elif target_probe_hybridization_probability_alignment_method == "bowtie":
-        # Specificity filter with Bowtie
-        target_probe_hybridization_probability_search_parameters = config[
-            "target_probe_hybridization_probability_bowtie_search_parameters"
-        ]
-        target_probe_hybridization_probability_hit_parameters = config[
-            "target_probe_hybridization_probability_bowtie_hit_parameters"
-        ]
-
-    target_probe_cross_hybridization_alignment_method = config[
```
[the remark this answers is not in the document]
Comment (jonas-hag) [reply oligo_designer_toolsuite/pipelines/_oligo_seq_probe_designer.py:1399]: return of `get_highly_abundant_kmer_sequences` should always be a list; also, missing defaults will be caught in the future by pydantic config validation

**Review by jonas-hag — APPROVED** [review]
